codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/BaseClientBuilderClass.java

```diff
@@ -241,9 +243,20 @@ private MethodSpec finalizeServiceConfigurationMethod() {
         if (StringUtils.isNotBlank(clientConfigClassName)) {
             ClassName clientConfigClass = ClassName.bestGuess(clientConfigClassName);
             builder.addCode("$1T.Builder c = (($1T) config.option($2T.SERVICE_CONFIGURATION)).toBuilder();" +
-                            "c.profileFile(c.profileFile() != null ? c.profileFile() : config.option($2T.PROFILE_FILE))" +
-                            " .profileName(c.profileName() != null ? c.profileName() : config.option($2T.PROFILE_NAME));",
+                            "c.profileFile(c.profileFile() != null ? c.profileFile() : config.option($2T.PROFILE_FILE));" +
+                            "c.profileName(c.profileName() != null ? c.profileName() : config.option($2T.PROFILE_NAME));",
                             clientConfigClass, SdkClientOption.class);
+
+            if (model.getCustomizationConfig().getServiceConfigHasDualstackConfig()) {
```

Comment:
Can this be replaced with a static public function in some util class and the implementation of if else can go there

```java
AwsConfigUtils.validateAndSetDualstackEnabledOption(ServiceConfiguration.Builder builder ,boolean dualEndpoint)
```

which can be called 

```java
AwsConfigUtils.validateAndSetDualstackEnabledOption(c,config.option($T.DUALSTACK_ENDPOINT_ENABLED))
```

Reply:
What would be the pros/cons?

Reply:
Pros : 
1. Readability in codegen code.
2. In future if we want to add some validation  or change the message of Validations , we need not change the codegen or its test cases.

Reply:
A few cons I see:
1. It adds a new protected API, which means we have a bigger backwards-compatibility surface area to worry about. As-is, the code is entirely within the service. We could get around that by finding a way to copy the implementation into every service module that needs it, but I think that would hurt codegen readability more than what we have here.
2. The service configurations do not share an interface, so we'd have to add a parent interface to the builder that allows reading and writing the dualstack endpoint... That one would be a new public API, which means even more backwards-compatibility surface area.

As for your pros:
1. I'm not entirely convinced it's much more readable to put into a separate function, since it's just 3 lines of code and it needs such a long function name to describe what it does.
2. It definitely makes it more testable in isolation, but we'd still have to test the fact that it is wired up in codegen, so we'd still have to test it indirectly like the test cases I added. If it was more than 3 lines, then I think there's a stronger argument for testability, since testing that it's wired up would be less work than actually testing every edge case within the function.

Reply:
Thanks for listing the details  👍 , good leanings for me